Noise mixer

mychiux413 opened the pull request:
## What does this PR do?
Hi,
This PR is an augmentation option to mix noise audio in process `RawAudioDataset.post_process()`,
sometimes we want to apply our ASR model just in some specific sound environments,
to make the model more adaptive to the target environment,
mixing the target environment sound into datasets while training should be better than `SpecAugment` to decrease `WER`.
 

Usage:
you can add the args below when execute `train.py` or `example/speech_recognition/infer.py`:
`--train-noise-dir`: str, A directory, put the soundfile readable files into it, the progress will automatically mix them into train datasets.
`--valid-noise-dir`: str, A directory, put the soundfile readable files into it, the progress will automatically mix them into valid datasets.
`--test-noise-dir`: str, A directory, put the soundfile readable files into it, the progress will automatically mix them into test datasets.
`--noise-min-snr-db`: float, default=3, The minimum signal-to-noise ratio between audio and noise.
`--noise-max-snr-db`: float, default=40, The maximum signal-to-noise ratio between audio and noise.

P.S. To test if the noise really mixed into datasets properly or not, we should hack into `RawAudioDataset.post_process` to dump the mixed feats out by `sf.write`, and listen it by our EARS.

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃
YES!

alexeib commented:
we have some internal impl based on https://github.com/facebookresearch/WavAugment, which i believe can also do additive noise. i have a dataset that can be used to wrap around audio dataset to add the noise, like this:

```
# Copyright (c) Facebook, Inc. and its affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

import numpy as np
import torch

from .. import BaseWrapperDataset


class ChainRunner(object):
    """
    Takes an instance of augment.EffectChain and applies it on pytorch tensors.
    """

    def __init__(self, chain):
        self.chain = chain

    def __call__(self, x):
        """
        x: torch.Tensor, (channels, length). Must be placed on CPU.
        """
        x = x.view(1, -1)
        src_info = {'channels': x.size(0),  # number of channels
                    'length': x.size(1),  # length of the sequence
                    'precision': 32,  # precision (16, 32 bits)
                    'rate': 16000.0,  # sampling rate
                    'bits_per_sample': 32}  # size of the sample

        target_info = {'channels': 1,
                       'length': x.size(1),
                       'precision': 32,
                       'rate': 16000.0,
                       'bits_per_sample': 32}

        y = self.chain.apply(x, src_info=src_info, target_info=target_info)

        # sox might misbehave sometimes by giving nan/inf if sequences are too short (or silent)
        # and the effect chain includes eg `pitch`
        if torch.isnan(y).any() or torch.isinf(y).any():
            return x
        return y.squeeze(0)


class AudioAugmentDataset(BaseWrapperDataset):
    def __init__(self, dataset):
        super().__init__(dataset)

        import augment
        effect_chain = augment.EffectChain() \
            .pitch("-q", self.random_pitch_shift) \
            .rate("-q", 16_000) \
            .reverb(50, 50, self.random_room_size).channels() \
            .time_dropout(max_seconds=50. / 1_000)
        # .additive_noise(noise_generator, snr=0, denormalize_noise=False) \

        self.runner = ChainRunner(effect_chain)

    def random_pitch_shift(self):
        return np.random.randint(-300, 300)

    def random_room_size(self):
        return np.random.randint(0, 100)

    def __getitem__(self, index):
        item = self.dataset[index]
        item["original_source"] = item["source"]
        item["source"] = self.runner(item["source"])

        return item
```

do you think it makes sense to rework your pr to use this kind of approach instead?

mychiux413 commented:
That would be great to introduce WavAugment into fairseq framework, and it has even more augment functions.

alexeib commented on fairseq/data/audio/raw_audio_dataset.py:
```diff
@@ -42,6 +49,13 @@ def __init__(
+        self.src_info = {"rate": sample_rate}
+        self.target_info = {"rate": sample_rate}
+        if wav_augment:
+            from fairseq.data.audio.wav_augment import chain_fatory
```
"fatory"?

mychiux413 replied:
fixed

alexeib commented on fairseq/data/audio/wav_augment.py:
```diff
@@ -0,0 +1,151 @@
+    return value
+
+
+def chain_fatory(commands) -> CustomEffectChain:
```
fatory?

mychiux413 replied:
fixed

alexeib reviewed (changes requested):
this is looking good! could you please move the augment functionality into a wrapper dataset (like in my previous comment) and then choose to wrap or not wrap a dataset in audio_pretraining task based on a flag

something like this:

```
        manifest = os.path.join(self.args.data, "{}.tsv".format(split))
        self.datasets[split] = FileAudioDataset(
            manifest,
            sample_rate=self.args.sample_rate,
            max_sample_size=self.args.max_sample_size,
            min_sample_size=self.args.max_sample_size,
            min_length=self.args.min_sample_size,
            pad=self.args.labels is not None or self.args.enable_padding,
            normalize=self.args.normalize,
            rescale=getattr(self.args, 'rescale', False),
            speakers=self.speakers,
        )

        if getattr(self.args, 'augment', False) and split not in self.args.valid_subset:
            self.datasets[split] = AudioAugmentDataset(self.datasets[split])
```

alexeib commented on fairseq/data/audio/raw_audio_dataset.py:
```diff
@@ -11,11 +11,17 @@
+    import soundfile as sf
+except ImportError as err:
+    print(err)
+    print('try: pip install soundfile')
```
this is going to get printed out for other people who dont use speech related functionality, since we read all datasets in data/__init_\_

mychiux413 replied:
fixed

alexeib commented on fairseq/data/audio/wav_augment.py:
```diff
@@ -0,0 +1,151 @@
+    from augment import EffectChain
+except ImportError as err:
+    err_msg = str(err)
+    err_msg += "\nTo apply the wav_augment, you should install the `facebookresearch/WavAugment` first.\n" + \
```
also this - maybe import in init()?

mychiux413 replied:
fixed

mychiux413 commented:
> this is looking good! could you please move the augment functionality into a wrapper dataset (like in my previous comment) and then choose to wrap or not wrap a dataset in audio_pretraining task based on a flag
> 
> something like this:
> 
> ```
>         manifest = os.path.join(self.args.data, "{}.tsv".format(split))
>         self.datasets[split] = FileAudioDataset(
>             manifest,
>             sample_rate=self.args.sample_rate,
>             max_sample_size=self.args.max_sample_size,
>             min_sample_size=self.args.max_sample_size,
>             min_length=self.args.min_sample_size,
>             pad=self.args.labels is not None or self.args.enable_padding,
>             normalize=self.args.normalize,
>         )
> 
>         if getattr(self.args, 'augment', False) and split not in self.args.valid_subset:
>             self.datasets[split] = AudioAugmentDataset(self.datasets[split])
> ```

I implemented the AudioAugmentDataset, but wrapping FileAudioDataset can make the `normalize` processing before `WavAugment`, I'm not sure whether it will affect the model while training with normalize.

alexeib commented:
i believe wavaugment expects audio data to be normalized, and also rescaled to [-1,1] (which --normalize currently doesnt do). when i was playing around with this i was doing something like this:

```
        if self.normalize or self.rescale:
            with torch.no_grad():
                feats = F.layer_norm(feats, feats.shape)
                if self.rescale:
                    feats = feats / feats.abs().max()
```
